<launch>
  <arg name="odometry_topic"/>

  <arg name="max_vel"/>
  <arg name="max_acc"/>

  <arg name="point_num"/>
  <arg name="point0_x"/>
  <arg name="point0_y"/>
  <arg name="point0_z"/>
  <arg name="point1_x"/>
  <arg name="point1_y"/>
  <arg name="point1_z"/>
  <arg name="point2_x"/>
  <arg name="point2_y"/>
  <arg name="point2_z"/>
  <arg name="point3_x"/>
  <arg name="point3_y"/>
  <arg name="point3_z"/>
  <arg name="point4_x"/>
  <arg name="point4_y"/>
  <arg name="point4_z"/>
  <arg name="point5_x"/>
  <arg name="point5_y"/>
  <arg name="point5_z"/>
  <arg name="point6_x"/>
  <arg name="point6_y"/>
  <arg name="point6_z"/>
  <arg name="point7_x"/>
  <arg name="point7_y"/>
  <arg name="point7_z"/>
  <arg name="point8_x"/>
  <arg name="point8_y"/>
  <arg name="point8_z"/>
  <arg name="point9_x"/>
  <arg name="point9_y"/>
  <arg name="point9_z"/>

  <arg name="flight_type"/>

  <!-- main node -->
  <node pkg="plan_manage" name="il_planner_node" type="il_planner_node" output="screen">
    <remap from="/odom_world" to="$(arg odometry_topic)"/>

    <!-- replanning method -->
    <param name="planner_node/planner" value="3" type="int"/>

    <!-- planning fsm -->
    <param name="fsm/flight_type" value="$(arg flight_type)" type="int"/>
    <param name="fsm/thresh_replan" value="1.5" type="double"/>
    <param name="fsm/thresh_no_replan" value="2.0" type="double"/>

    <param name="fsm/waypoint_num" value="$(arg point_num)" type="int"/>
    <param name="fsm/waypoint0_x" value="$(arg point0_x)" type="double"/>
    <param name="fsm/waypoint0_y" value="$(arg point0_y)" type="double"/>
    <param name="fsm/waypoint0_z" value="$(arg point0_z)" type="double"/>
    <param name="fsm/waypoint1_x" value="$(arg point1_x)" type="double"/>
    <param name="fsm/waypoint1_y" value="$(arg point1_y)" type="double"/>
    <param name="fsm/waypoint1_z" value="$(arg point1_z)" type="double"/>
    <param name="fsm/waypoint2_x" value="$(arg point2_x)" type="double"/>
    <param name="fsm/waypoint2_y" value="$(arg point2_y)" type="double"/>
    <param name="fsm/waypoint2_z" value="$(arg point2_z)" type="double"/>
    <param name="fsm/waypoint3_x" value="$(arg point3_x)" type="double"/>
    <param name="fsm/waypoint3_y" value="$(arg point3_y)" type="double"/>
    <param name="fsm/waypoint3_z" value="$(arg point3_z)" type="double"/>
    <param name="fsm/waypoint4_x" value="$(arg point4_x)" type="double"/>
    <param name="fsm/waypoint4_y" value="$(arg point4_y)" type="double"/>
    <param name="fsm/waypoint4_z" value="$(arg point4_z)" type="double"/>
    <param name="fsm/waypoint5_x" value="$(arg point5_x)" type="double"/>
    <param name="fsm/waypoint5_y" value="$(arg point5_y)" type="double"/>
    <param name="fsm/waypoint5_z" value="$(arg point5_z)" type="double"/>
    <param name="fsm/waypoint6_x" value="$(arg point6_x)" type="double"/>
    <param name="fsm/waypoint6_y" value="$(arg point6_y)" type="double"/>
    <param name="fsm/waypoint6_z" value="$(arg point6_z)" type="double"/>
    <param name="fsm/waypoint7_x" value="$(arg point7_x)" type="double"/>
    <param name="fsm/waypoint7_y" value="$(arg point7_y)" type="double"/>
    <param name="fsm/waypoint7_z" value="$(arg point7_z)" type="double"/>
    <param name="fsm/waypoint8_x" value="$(arg point8_x)" type="double"/>
    <param name="fsm/waypoint8_y" value="$(arg point8_y)" type="double"/>
    <param name="fsm/waypoint8_z" value="$(arg point8_z)" type="double"/>
    <param name="fsm/waypoint9_x" value="$(arg point9_x)" type="double"/>
    <param name="fsm/waypoint9_y" value="$(arg point9_y)" type="double"/>
    <param name="fsm/waypoint9_z" value="$(arg point9_z)" type="double"/>

    <!-- tube rrt* params -->
    <param name="planParam/max_radius"      value="1.5"  type="double"/>
    <param name="planParam/safety_margin"   value="0.75" type="double"/>
    <param name="planParam/search_margin"   value="0.25" type="double"/>

  <!-- planner manager -->
    <param name="ILParam/vmax" value="5.0" type="double"/>
    <param name="ILParam/vmin" value="1.0" type="double"/>
    <param name="ILParam/goalth" value="1.0" type="double"/>
    <param name="ILParam/xth" value="0.6" type="double"/>
    <param name="ILParam/kp_law" value="1.0" type="double"/>
    <param name="ILParam/kp_vl" value="1.5" type="double"/>
    <param name="ILParam/kd_law" value="0.6" type="double"/>
    <param name="ILParam/kp_path" value="3.2" type="double"/>
    <param name="ILParam/kd_path" value="4.5" type="double"/>
    <param name="ILParam/iteration" value="30" type="int"/>
    <param name="ILParam/dynamic" value="3.0" type="double"/>
    <param name="ILParam/amax" value="6.0" type="double"/>


    <param name="manager/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="manager/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="manager/max_jerk" value="4" type="double"/>
    <param name="manager/control_points_distance" value="0.4" type="double"/>
    <param name="manager/feasibility_tolerance" value="0.05" type="double"/>
    <param name="manager/planning_horizon" value="7.5" type="double"/>
    <param name="manager/use_geometric_path" value="false" type="bool"/>
    <param name="manager/use_kinodynamic_path" value="false" type="bool"/>
    <param name="manager/use_topo_path" value="false" type="bool"/>
    <param name="manager/use_tube_path" value="true" type="bool"/>
    <param name="manager/use_optimization" value="true" type="bool"/>
    <param name="manager/use_active_perception" value="true" type="bool"/>
    <param name="manager/min_time" value="true" type="bool"/>

  <!-- kinodynamic path searching -->
    <param name="search/max_tau" value="0.6" type="double"/>
    <param name="search/init_max_tau" value="0.8" type="double"/>
    <param name="search/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="search/max_acc" value="$(arg max_acc)" type="double"/>
    <param name="search/w_time" value="10.0" type="double"/>
    <param name="search/horizon" value="7.0" type="double"/>
    <param name="search/lambda_heu" value="5.0" type="double"/>
    <param name="search/resolution_astar" value="0.1" type="double"/>
    <param name="search/time_resolution" value="0.8" type="double"/>
    <param name="search/margin" value="0.2" type="double"/>
    <param name="search/allocate_num" value="100000" type="int"/>
    <param name="search/check_num" value="5" type="int"/>

  <!-- trajectory optimization -->
    <param name="optimization/lambda1" value="10.0" type="double"/>
    <param name="optimization/lambda2" value="5.0" type="double"/>
    <param name="optimization/lambda3" value="0.00001" type="double"/>
    <param name="optimization/lambda4" value="0.01" type="double"/>
    <param name="optimization/lambda5" value="1.5" type="double"/>
    <param name="optimization/lambda6" value="10.0" type="double"/>
    <param name="optimization/lambda7" value="100.0" type="double"/>
    <param name="optimization/dist0" value="0.5" type="double"/>
    <param name="optimization/dist1" value="1.0" type="double"/>
    <param name="optimization/max_vel" value="$(arg max_vel)" type="double"/>
    <param name="optimization/max_acc" value="$(arg max_acc)" type="double"/>

    <param name="optimization/algorithm1" value="40" type="int"/>
    <param name="optimization/algorithm2" value="40" type="int"/>
    <param name="optimization/max_iteration_num1" value="200" type="int"/>
    <param name="optimization/max_iteration_num2" value="200" type="int"/>
    <param name="optimization/max_iteration_num3" value="200" type="int"/>
    <param name="optimization/max_iteration_num4" value="200" type="int"/>
    <param name="optimization/max_iteration_time1" value="0.01" type="double"/>
    <param name="optimization/max_iteration_time2" value="0.01" type="double"/>
    <param name="optimization/max_iteration_time3" value="0.01" type="double"/>
    <param name="optimization/max_iteration_time4" value="0.01" type="double"/>
    <param name="optimization/order" value="3" type="int"/>

  </node>

</launch>
